Add `message_ids` parameter to /messages API endpoint.

amanagr opened the pull request:
discussion: https://chat.zulip.org/#narrow/stream/378-api-design/topic/Message.20matches.20narrow

amanagr committed:
message_events: Fix property_value not being passed.

This was not passed by mistake.

amanagr committed:
message_events: Use the new message_ids param of /messages.

This helps us better update the views where we don't have the
messages that were updated cached or the filter cannot be
applied locally.

Tested using browser breakpoints for search `zulip` in starred narrow
for `!filter.can_apply_locally()` and setting message_fetch
constants to single digit for `messages_to_fetch.length > 0`.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -6618,6 +6621,18 @@ paths:
+        - name: message_ids
+          in: query
+          description: |
+            A list of message IDs to fetch. If provided, `anchor`, `num_before`,
+            and `num_after` are ignored.
```
If `message_ids` is provided, then those other parameters should not just be ignored but be an error.

Also `include_anchor` should get the same treatment as those three — it's an error to provide it at the same time as `message_ids`.

timabbott replied:
The code seems to do that, so this documentation is wrong.

amanagr replied:
Oops, updated that comment to explain things in more detail.

timabbott commented on zerver/views/message_fetch.py:
```diff
@@ -106,19 +106,40 @@ def get_messages_backend(
+    if message_ids_to_fetch is None:
+        message_ids_to_fetch = []
+    # User requested a list of message ids to fetch.
+    is_message_ids_pre_specified = len(message_ids_to_fetch) > 0
```
Might be good to integrate this with the `message_ids_to_fetch` None check above; depending on the answer to https://chat.zulip.org/#narrow/stream/378-api-design/topic/Message.20matches.20narrow/near/1951550, we might want to handle no parameter vs. empty parameter differently. 

I think we can also do the following names:
- `message_ids_to_fetch` => `client_requested_message_ids` -- makes really clear these aren't just some internal value partway through the normal fetching process.
- `is_message_ids_pre_specified` => Do we need this, vs. just checking `if client_requested_message_ids is not None` and setting the value for that properly?

amanagr replied:
Removed `is_message_ids_pre_specified` and did the rename.

timabbott commented on zerver/views/message_fetch.py:
```diff
@@ -106,19 +106,40 @@ def get_messages_backend(
+        num_before or num_after or anchor_val is not None or use_first_unread_anchor_val
+    ) and is_message_ids_pre_specified:
+        raise JsonableError(
+            _("`message_ids` are mutually exclusive with `num_before`, `num_after` and `anchor`")
```
You can use `IncompatibleParametersError` to avoid extra overly specific translated strings.

amanagr replied:
Done

timabbott commented on zerver/views/message_fetch.py:
```diff
@@ -212,6 +233,7 @@ def get_messages_backend(
             include_anchor=include_anchor,
             num_before=num_before,
             num_after=num_after,
+            message_ids=message_ids_to_fetch,
```
I'd prefer to keep the parameter name unchanged in this API boundary, to keep the code easier to trace.

amanagr replied:
Okay, used `client_requested_message_ids`.

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -874,7 +875,8 @@ def test_generate_and_render_curl_example_with_excludes(self) -> None:
             '    --data-urlencode \'narrow=[{"operand": "Denmark", "operator": "channel"}]\' \\',
-            "    --data-urlencode use_first_unread_anchor=true",
+            "    --data-urlencode use_first_unread_anchor=true \\",
+            "    --data-urlencode 'message_ids=[1, 2, 3]'",
```
Generally, we want these tests to be sample of normal requests to make, and I think that means we don't want to add this new rare-use parameter here. Also this reveals that `use_first_unread_anchor` needs to be incompatible with the new parameter. 

(You might also want a prep commit to rework these examples to use the modern API; `use_first_unread_anchor` is deprecated, right?)

amanagr replied:
Added a prep commit to remove usage of `use_first_unread_anchor` and added `message_ids` to the excluded parameteres list.

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -1389,47 +1391,68 @@ def fetch_messages(
+            anchor=None,
+            include_history=include_history,
+            is_search=is_search,
+        )
```
Where are you doing the security checks that this user has access to the message IDs specified? We definitely need tests for that. It might be that #29380 is a complete answer to that, which @prakhar1144 is currently working on getting finished, but we still need a robust test suite in case refactoring breaks the logic.

amanagr replied:
Added a comment for it there and some tests.

timabbott commented on zerver/views/message_fetch.py:
```diff
@@ -106,19 +106,40 @@ def get_messages_backend(
     narrow = update_narrow_terms_containing_with_operator(realm, maybe_user_profile, narrow)
-    if num_before + num_after > MAX_MESSAGES_PER_FETCH:
+
+    if num_before + num_after + len(message_ids_to_fetch) > MAX_MESSAGES_PER_FETCH:
```
Cool, this is a nice way to do it. Make sure you have a test covering this.

amanagr replied:
No longer using this after the naming changes.

amanagr committed:
message_fetch: Rename message_ids variable for clarity.

Added `result_` prefix to differentiate it from upcoming `message_ids`
parameter to the API request. Also, this is final `message_ids` that
we will fetch the messages for. So, a `result` prefix makes sense here.

amanagr committed:
test_openapi: Remove use_first_unread_anchor from curl examples.

amanagr committed:
message_fetch: Add message_ids parameter to /messages request.

This allows us to fetch messages for a list of message ids in a
single request.

amanagr commented on zerver/views/message_fetch.py:
```diff
@@ -226,33 +248,33 @@ def get_messages_backend(
         # bulk-fetch rendered message dicts from remote cache using the
         # 'messages' list.
-        message_ids: list[int] = []
+        filtered_message_ids: list[int] = []
```
Added a prep commit here.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -6648,6 +6651,23 @@ paths:
+            A list of message IDs to fetch. Server will return messages matching
+            the narrow and the message IDs provided.
+
+            It is mutually exclusive with some properties like `anchor`, `num_before`
+            and `num_after`, i.e., if `message_ids` is provided, then `anchor`,
```
This is the reference documentation, so it should be definitive about exactly which properties it's referring to here.

amanagr replied:
Updated to remove `some properties like`.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -6666,18 +6690,24 @@ paths:
                           `use_first_unread_anchor` is `true`).
+
+                          Only present if `message_ids` is not provided.
                       found_newest:
                         type: boolean
                         description: |
                           Whether the server promises that the `messages` list includes the very
                           newest messages matching the narrow (used by clients that paginate their
                           requests to decide whether there may be more messages to fetch).
+
+                          Only present if `message_ids` is not provided.
```
In the `#api design` thread, the proposal was for `found_newest` and `found_oldest` to continue to be always present:
https://chat.zulip.org/#narrow/stream/378-api-design/topic/Message.20matches.20narrow/near/1951604

amanagr commented:
This is ready for another review.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -6648,6 +6651,23 @@ paths:
+            A list of message IDs to fetch. Server will return messages matching
+            the narrow and the message IDs provided.
+
+            It is mutually exclusive with `anchor`, `num_before`
+            and `num_after`, i.e., if `message_ids` is provided, then `anchor`,
```
Bump from https://github.com/zulip/zulip/pull/31787#discussion_r1781463688 re `include_anchor`.

amanagr replied:
Sorry, I missed it somehow. Added to `include_anchor` to the mutually excluded list too.

timabbott commented on zerver/views/message_fetch.py:
```diff
@@ -226,33 +226,33 @@ def get_messages_backend(
             for row in rows:
                 message_id = row[0]
-                message_ids.append(message_id)
+                filtered_message_ids.append(message_id)
```
I think `result_message_ids` would be a clearer name?

amanagr replied:
Yeah, replaced the prep commit with that name and added a comment.

timabbott commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,13 @@ format used by the Zulip server that they are interacting with.
+* [`GET /messages`](/api/get-message): Added a parameter `message_ids` to
+  fetch messages provided a list of message ids. `message_ids` is mutually
+  exclusive to `anchor`, `num_after` and `num_before`. Note that this makes
+  all the parameters of GET request to `/messages` API endpoint optional.
```
I would write for the prose:

> Added a new `message_ids` parameter, as an alternative method of specifying which messages to fetch.

amanagr replied:
Changed it here and added a **changes** entry to `get` description.

timabbott commented on web/src/message_events.js:
```diff
@@ -118,21 +118,44 @@ export function update_views_filtered_on_message_property(
         // In case of multiple messages, we just rerender the view
         // since it is likely to use similar amount of resources to
         // fetching the messages and rerendering the view.
```
This comment is no longer accurate.

amanagr replied:
Yeah, removed it.

timabbott commented on web/src/message_events.js:
```diff
@@ -118,21 +118,44 @@ export function update_views_filtered_on_message_property(
-                    message_helper.process_new_message(data.message);
+                    for (const raw_message of data.messages) {
+                        message_helper.process_new_message(raw_message);
+                    }
```
Since there's network delay involved, we probably still need to consider that we received a copy of the message in between calculating `messages_to_fetch` and this returning. I think `message_helper.process_new_message` handles that, but maybe we should note that possibility in a comment here.

amanagr replied:
Added a comment.

```
// `messages_to_fetch` might already be cached locally when
// we reach here but `message_helper.process_new_message`
// already handles that case.
```

timabbott commented on web/src/message_events.js:
```diff
@@ -118,21 +118,44 @@ export function update_views_filtered_on_message_property(
+                    msg_list.data.remove([...messages_to_remove]);
+                    msg_list.data.add_messages(messages_to_add);
+                    msg_list.rerender();
+                },
```
The structure here, where we do this `channel.get` call inside the `message_lists.all_rendered_message_lists` loop, means we could in theory do multiple of this query in a way that's somewhat inefficient.

I also note that I think we need to do this logic for non-rendered cached message_list_data objects. So we might need, perhaps in a follow-up PR, to think about reworking the loop structure to first collect the messages to refetch into a single server query? Maybe I'm missing something.

amanagr replied:
I agree we need to think about optimizing it but will be more to think about when adding the logic for cached message list data objects.

timabbott commented on zerver/views/message_fetch.py:
```diff
@@ -106,19 +110,40 @@ def get_messages_backend(
+        num_before or num_after or anchor_val is not None or use_first_unread_anchor_val
+    ) and client_requested_message_ids is not None:
+        raise IncompatibleParametersError(
+            ["num_before", "num_after", "anchor", "message_ids", "use_first_unread_anchor_val"]
```
I think this should be `use_first_unread_anchor`; the `_val` is a local variable detail, not part of the API.

amanagr replied:
Oops! Fixed.

timabbott commented on zerver/tests/test_message_fetch.py:
```diff
@@ -2085,6 +2085,97 @@ def test_unauthenticated_get_messages(self) -> None:
+            self.unsubscribe(hamlet, stream_name)
+            self.subscribe(iago, stream_name)
+            message_ids.append(self.send_stream_message(self.example_user("iago"), stream_name))
+        self.logout()
```
This seems a bit contorted, switching user accounts so many times. Generally we use `self.api_post` to avoid needing to switch users for individual operations ... but also I don't see how `hamlet` would have been subscribed to those channels in the first place, if they're newly created by Iago?

amanagr replied:
Okay, removed the hamlet calls to unsubscribe here. I thought the test might be more idempotent with those calls.

timabbott replied:
Tests don't need to be idempotent from a database state standpoint; in Django, the test runs in a transaction that is rolled back at the end of the test.

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -1389,47 +1392,68 @@ def fetch_messages(
+            history_limited=False,
+            anchor=None,
+            include_history=include_history,
+            is_search=is_search,
```
Returning early here looks to skip checking `first_visible_message_id`.

timabbott replied:
I'll add code for this here, but probably some refactoring could make it a bit cleaner.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -6506,6 +6506,12 @@ paths:
+        of specifying which messages to fetch. `message_ids` is mutually
+        exclusive to `anchor`, `include_anchor`, `num_after`, `num_before` and
+        `use_first_unread_anchor` (deprecated). Note that this makes all the
+        parameters GET request to `/messages` API endpoint optional.
```
Changes entries always need to mention the API feature level. I ended up reworking the docs significantly in this review pass; here's the diff:

``` diff
diff --git a/zerver/lib/narrow.py b/zerver/lib/narrow.py
index 01f35911c4..ad6df665c3 100644
--- a/zerver/lib/narrow.py
+++ b/zerver/lib/narrow.py
@@ -1442,8 +1442,12 @@ def fetch_messages(
 
     if client_requested_message_ids is not None:
         # We don't need to do any post-processing in this case.
+        if first_visible_message_id > 0:
+            visible_rows: Sequence[MessageRowT] = [r for r in rows if r[0] >= first_visible_message_id]
+        else:
+            visible_rows = rows
         return FetchedMessages(
-            rows=rows,
+            rows=visible_rows,
             found_anchor=False,
             found_newest=False,
             found_oldest=False,
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 85286976bf..e78485ec37 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -6479,39 +6479,37 @@ paths:
       summary: Get messages
       tags: ["messages"]
       description: |
-        Fetch user's message history from a Zulip server.
+        This endpoint is the primary way to fetch a messages. It is used by all official
+        Zulip clients (e.g. the web, desktop, mobile, and terminal clients) as well as
+        many bots, API clients, backup scripts, etc.
 
-        This endpoint is the primary way to fetch a user's message history
-        from a Zulip server. It is useful both for Zulip clients (e.g. the
-        web, desktop, mobile, and terminal clients) as well as bots, API
-        clients, backup scripts, etc.
+        Most queries will specify a [narrow filter](/api/get-messages#parameter-narrow),
+        to fetch the messages matching any supported [search
+        query](/help/search-for-messages). If not specified, it will return messages
+        corresponding to the user's [combined feed](/help/combined-feed). There are two
+        ways to specify which messages matching the narrow filter to fetch:
+
+        - A range of messages, described by an `anchor` message ID (or a string-format
+          specification of how the server should computer an anchor to use) and a maximum
+          number of messages in each direction from that anchor.
+
+        - A rarely used variant (`message_ids`) where the client specifies the message IDs
+          to fetch.
+
+        The server returns the matching messages, sorted by message ID, as well as some
+        metadata that makes it easy for a client to determine whether there are more
+        messages matching the query that were not returned due to the `num_before` and
+        `num_after` limits.
 
         Note that a user's message history does not contain messages sent to
         channels before they [subscribe](/api/subscribe), and newly created
         bot users are not usually subscribed to any channels.
 
-        By specifying a [narrow filter](/api/get-messages#parameter-narrow),
-        you can use this endpoint to fetch the messages matching any search
-        query that is supported by Zulip's powerful full-text search backend.
-
-        In either case, you specify an `anchor` message (or ask the server to
-        calculate the first unread message for you and use that as the
-        anchor), as well as a number of messages before and after the anchor
-        message. The server returns those messages, sorted by message ID, as
-        well as some metadata that makes it easy for a client to determine
-        whether there are more messages matching the query that were not
-        returned due to the `num_before` and `num_after` limits.
-
-        We recommend setting `num_before` and `num_after` to no more than 1000, to
-        avoid generating very large HTTP responses. A maximum of 5000 messages
-        can be obtained per request; attempting to exceed this will result in an
-        error.
-
-        **Changes**: Added a new `message_ids` parameter, as an alternative method
-        of specifying which messages to fetch. `message_ids` is mutually
-        exclusive to `anchor`, `include_anchor`, `num_after`, `num_before` and
-        `use_first_unread_anchor` (deprecated). Note that this makes all the
-        parameters GET request to `/messages` API endpoint optional.
+        We recommend requesting at most 1000 messages in a batch, to avoid generating very
+        large HTTP responses. A maximum of 5000 messages can be obtained per request;
+        attempting to exceed this will result in an error.
+
+        **Changes**: The `message_ids` option is new in Zulip 10.0 (feature level 300).
       x-curl-examples-parameters:
         oneOf:
           - type: exclude
@@ -6660,13 +6658,16 @@ paths:
         - name: message_ids
           in: query
           description: |
-            A list of message IDs to fetch. Server will return messages matching
-            the narrow and the message IDs provided.
+            A list of message IDs to fetch. The server will return messages corresponding to the
+            subset of the requested message IDs that exist and the current user has access to,
+            potentially filtered by the narrow (if that parameter is provided).
+
+            It is an error to pass this parameter as well as any of the parameters involved in
+            specifying a range of messages: `anchor`, `include_anchor`, `use_first_unread_anchor`,
+            `num_before`, and `num_after`.
 
-            It is mutually exclusive with `anchor`, `include_anchor`, `num_before`,
-            and `num_after`, i.e., if `message_ids` is provided, then `anchor`,
-            `include_anchor`, `num_before` and `num_after` should not be provided
-            otherwise it will result in an error.
+            **Changes**: New in Zulip 10.0 (feature level 300). Previously, there was
+            no way to request a specific set of messages IDs.
           content:
             application/json:
               schema:
```

timabbott commented:
OK I made the tweaks noted above and marked this to merge once CI passes. Thanks for all the work on this @amanagr!

timabbott commented:
Guess it needs test coverage for the new visibility logic, or maybe a refactor to make that a shared code path with the existing filtering of that form, which might be better; I don't actually want to duplicate that conditional.

amanagr commented:
Okay, I will take care of it.

timabbott commented:
Merged, thanks @amanagr!
